lib/receive.dart

```diff
@@ -202,78 +158,157 @@ class MaybeInputsSeen extends FfiV2MaybeInputsSeen {
+          replacementOutputs: replacementOutputs, drainScript: drainScript);
+      return WantsOutputs._(field0: res.field0);
     } on error.PayjoinError catch (e) {
       throw mapPayjoinError(e);
     }
   }
 
   @override
-  Future<void> contributeWitnessInput(
-      {required TxOut txo, required OutPoint outpoint, hint}) {
+  Future<WantsOutputs> substituteReceiverScript(
+      {required FfiScript outputScript}) async {
     try {
-      return super.contributeWitnessInput(txo: txo, outpoint: outpoint);
+      final res =
+          await super.substituteReceiverScript(outputScript: outputScript);
+      return WantsOutputs._(field0: res.field0);
     } on error.PayjoinError catch (e) {
       throw mapPayjoinError(e);
     }
   }
 
-  /// Select receiver input such that the common.dart avoids surveillance. Return the input chosen that has been applied to the Proposal.
+  @override
+  Future<WantsInputs> commitOutputs() async {
+    try {
+      final res = await super.commitOutputs();
+      return WantsInputs._(field0: res.field0);
+    } on error.PayjoinError catch (e) {
+      throw mapPayjoinError(e);
+    }
+  }
+}
+
+class WantsInputs extends FfiWantsInputs {
+  WantsInputs._({required super.field0});
+
+  /// Select receiver input such that the payjoin avoids surveillance.
+  /// Return the input chosen that has been applied to the Proposal.
   ///
-  /// Proper coin selection allows common.dart to resemble ordinary transactions. To ensure the resemblance, a number of heuristics must be avoided.
+  /// Proper coin selection allows payjoin to resemble ordinary transactions.
+  /// To ensure the resemblance, a number of heuristics must be avoided.
   ///
-  /// UIH “Unnecessary input heuristic” is one class of them to avoid. We define UIH1 and UIH2 according to the BlockSci practice BlockSci UIH1 and UIH2:
+  /// UIH "Unnecessary input heuristic" is one class of them to avoid. We define
+  /// UIH1 and UIH2 according to the BlockSci practice
+  /// BlockSci UIH1 and UIH2:
+  /// if min(out) < min(in) then UIH1 else UIH2
+  /// https://eprint.iacr.org/2022/589.pdf
   @override
-  Future<OutPoint> tryPreservingPrivacy(
-      {required Map<BigInt, OutPoint> candidateInputs, hint}) {
+  Future<InputPair> tryPreservingPrivacy(
+      {required List<FfiInputPair> candidateInputs}) async {
     try {
-      return super.tryPreservingPrivacy(candidateInputs: candidateInputs);
+      final res =
+          await super.tryPreservingPrivacy(candidateInputs: candidateInputs);
+      return InputPair._(field0: res.field0);
     } on error.PayjoinError catch (e) {
       throw mapPayjoinError(e);
     }
   }
 
   @override
-  Future<PayjoinProposal> finalizeProposal(
-      {required FutureOr<String> Function(String p1) processPsbt,
-      BigInt? minFeeRateSatPerVb,
-      hint}) async {
+  Future<WantsInputs> contributeInputs(
+      {required List<FfiInputPair> replacementInputs}) async {
```

Comment:
Same to better use `InputPair` instead of internal `FfiInputPair`.

Reply:
because the internal super.contributeInputs takes a List<FfiInputPair>, using List<InputPair> isn't a valid override

however, I have no problem passing List<InputPair> to the function in the example because InputPair extends FfiInputPair (see: https://github.com/LtbLightning/payjoin-flutter/blob/0a6aa08caf2448da55af844c1c5587a809f32cb6/example/lib/screens/home.dart#L581-L604)

Reply:
Solved in '[fix: exported internal ffi classes as parameter types](https://github.com/LtbLightning/payjoin-flutter/pull/38/commits/4b59d0ebc36f813491e823638bb57a6338f51f92)